File: aspnet/Tide.Asgard.AspNetCore/Tide.Asgard.Scheduler/Expression/Tokenizer.cs
// Copyright (c) Tide Foundation Limited. All rights reserved.
// Licensed under the Tide Community Open Code License. See LICENSE in the project root.

namespace Tide.Asgard.Scheduler.Expression;

// Character offset into the original expression, used for error reporting.
public readonly record struct Token(string Text, int Offset);

public static class Tokenizer
{
	// The grammar is whitespace separated words, so tokenizing is a split that
	// keeps offsets. Case handling is left to the parser because timezone ids
	// and instants are case sensitive.
	public static List<Token> Tokenize(string input)
	{
		var tokens = new List<Token>();
		var i = 0;

		while (i < input.Length)
		{
			while (i < input.Length && char.IsWhiteSpace(input[i])) i++;
			if (i >= input.Length) break;

			var start = i;
			while (i < input.Length && !char.IsWhiteSpace(input[i])) i++;
			tokens.Add(new Token(input[start..i], start));
		}

		return tokens;
	}
}

File: src/scheduler/expression/Tokenizer.ts
// Copyright (c) Tide Foundation Limited. All rights reserved.
// Licensed under the Tide Community Open Code License. See LICENSE in the project root.

export interface Token {
    readonly text: string;
    // Character offset into the original expression, used for error reporting.
    readonly offset: number;
}

// The grammar is whitespace separated words, so tokenizing is a split that
// keeps offsets. Case is normalised here; values that are case sensitive
// (timezone ids and instants) are re-read from the raw text by the parser.
export function tokenize(input: string): Token[] {
    const tokens: Token[] = [];
    let i = 0;

    while (i < input.length) {
        while (i < input.length && isSpace(input[i])) i++;
        if (i >= input.length) break;

        const start = i;
        while (i < input.length && !isSpace(input[i])) i++;
        tokens.push({ text: input.slice(start, i), offset: start });
    }

    return tokens;
}

function isSpace(ch: string): boolean {
    return ch === " " || ch === "\t" || ch === "\n" || ch === "\r";
}
